Fix protocol assert_conforms_to, relax NumericalFlux requirements for evolution DG transition

nilsvu opened the pull request:
## Proposed changes

- Fix a bug in `tt::assert_conforms_to` that made compilers optimize out an unused type alias
- Relax the requirements in the `NumericalFlux` protocol because the evolution DG schemes don't follow the interface documented in the protocol anymore.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Fix protocol assert_conforms_to

kidder commented on src/Utilities/ProtocolHelpers.hpp:
```diff
@@ -60,7 +60,11 @@ struct AssertConformsToImpl : std::true_type {
       "Have you forgotten to (publicly) inherit the type from "
       "tt::ConformsTo<Protocol>?");
-  using test = typename Protocol::template test<ConformingType>;
+  // The `decltype` makes sure the conformance test is not optimized out by
```
I am surprised if this ever worked as a the type alias you deleted does not require a definition of `Protocol::test`. I would change the comment to `// Implicitly instantiate Protocol::test in order to test conformance`

kidder reviewed (changes requested):
All the compilers are complaining (I assume about something not conforming to a protocol, but the error messages are not immediately enlightening...)

nilsvu committed:
Relax NumericalFlux protocol for evolution transition

The evolution DG schemes don't follow the interface documented in the
protocol anymore, so we relax the conditions to make the transition easier.
